### PR TITLE
Refactor (Fisher)LDA solvers into separate class

### PR DESCRIPTION
**Pull request by micmn**
This includes:
- refactor LDA binary classification and dimension reduction solvers in two separate classes: `LDASolver`, centers data w.r.t. class, compute within cov matrix (shared by bin. class. and dim. red. llt/eigen solvers); `LDACanVarSolver`, inherits from `LDASolver` and computes the solution through canonical variates algorithm (shared by bin class. and dim. red. canvar solvers)
- fix bias calculation in LDA #3886 
- option to compute SVD by bidiagonalization divide and conquer algorithm (with Eigen >= 3.3)
- MulticlassLabels constructor from BinaryLabels (I know there is already some work going on labels, so let me know if there is a better way)
- LDA benchmarks: https://gist.github.com/micmn/d4b51bf39676621afca563b55bdb9a98 :dancer:

Certainly some docs are missing, I'll double check.

### CODE REVIEW

**Comment by vigsterkr** [comment]
@micmn great! now the only thing that's left is to update the shogun-data repo with the new integration test data for binary lda (this is why all the jobs on travis failed). lemme know if you need help how to do it

File: src/shogun/classifier/LDA.cpp [hunk 1]
```diff
@@ -82,114 +83,81 @@ bool CLDA::train_machine(CFeatures *data)
...
-	// keep just the first dimension to do binary classification
-	lda->fit(data, multiclass_labels.get(), 1);
-	auto m = lda->get_transformation_matrix();
+	// change the sign of w if needed to get the correct labels
+	float64_t sign = (m_pos > m_neg) ? 1 : -1;
 
-	SGVector<float64_t> w(m);
+	SGVector<float64_t> w(dense_feat->get_num_features());
+	// copy w_st into w
+	for (index_t i = 0; i < w.size(); ++i)
```
Comment (vigsterkr) [inline src/shogun/classifier/LDA.cpp]: `linalg::scale(w_st, w, sign)` ?

File: src/shogun/classifier/LDA.cpp [hunk 2]
```diff
@@ -82,114 +83,81 @@ bool CLDA::train_machine(CFeatures *data)
...
-
-	// mean pos
-	for (auto i : idx_pos)
-		linalg::add_col_vec(matrix, i, mean_pos, mean_pos);
-	linalg::scale(mean_pos, mean_pos, 1 / (ST)idx_pos.size());
-
-	// get m(+ve) - mean(+ve)
-	for (auto i : idx_pos)
-		linalg::add_col_vec(matrix, i, mean_pos, matrix, (ST)1, (ST)-1);
+	auto dense_feat = static_cast<CDenseFeatures<ST>*>(features);
```
Comment (vigsterkr) [inline src/shogun/classifier/LDA.cpp]: we could add here a test or `REQUIRED` for checking that
```
REQUIRED(features->get_feature_class() == C_DENSE, "LDA only works with dense features")
```

File: src/shogun/solver/LDASolver.h [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+		linalg::zero(m_class_mean[i]);
+	}
+	m_mean = SGVector<T>(data.num_rows);
+	linalg::zero(m_mean);
+
+	// calculate the total mean and the classes' mean.
+	for (index_t i = 0; i < data.num_cols; i++)
+	{
+		index_t c = (index_t)m_labels->get_label(i);
+		m_class_count[c]++;
```
Comment (vigsterkr) [inline src/shogun/solver/LDASolver.h]: in this case it could be prefixed:
++m_class_count[c];

File: src/shogun/classifier/LDA.cpp [hunk 4]
```diff
@@ -11,50 +11,52 @@
...
+#include <shogun/solver/LDASolver.h>
+#include <shogun/solver/LDACanVarSolver.h>
 #include <vector>
 
 using namespace Eigen;
 using namespace shogun;
 
-CLDA::CLDA(float64_t gamma, ELDAMethod method)
-	:CLinearMachine()
+CLDA::CLDA(float64_t gamma, ELDAMethod method, bool accurate)
```
Comment (karlnapf) [inline src/shogun/classifier/LDA.cpp]: accurate what?

Comment (karlnapf) [inline src/shogun/classifier/LDA.cpp]: maybe a better name? that explains what is accurate?

File: src/shogun/classifier/LDA.cpp [hunk 5]
```diff
@@ -66,8 +68,7 @@ bool CLDA::train_machine(CFeatures *data)
 	REQUIRE(m_labels, "Labels for the given features are not specified!\n")
 	REQUIRE(
 	    m_labels->get_label_type() == LT_BINARY,
-	    "The labels should of type"
-	    " CBinaryLabels! you provided %s \n",
+	    "The labels should of type CBinaryLabels! you provided %s \n",
```
Comment (karlnapf) [inline src/shogun/classifier/LDA.cpp]: You with a capital Y
Also, better to ommit "you" in error messages

File: src/shogun/classifier/LDA.cpp [hunk 6]
```diff
@@ -82,114 +83,81 @@ bool CLDA::train_machine(CFeatures *data)
...
-	std::unique_ptr<CMulticlassLabels> multiclass_labels(
-	    new CMulticlassLabels(m_labels->get_num_labels()));
+	auto dense_feat = static_cast<CDenseFeatures<ST>*>(features);
+
+	// keep just one dimension to do binary classification
+	const index_t projection_dim = 1;
+	auto solver = std::unique_ptr<CLDACanVarSolver<ST> >(
+		new CLDACanVarSolver<ST>(
+			dense_feat,
+			new CMulticlassLabels(static_cast<CBinaryLabels*>(m_labels)),
```
Comment (karlnapf) [inline src/shogun/classifier/LDA.cpp]: is this all safe?

File: src/shogun/classifier/LDA.cpp [hunk 7]
```diff
@@ -82,114 +83,81 @@ bool CLDA::train_machine(CFeatures *data)
...
+	auto dense_feat = static_cast<CDenseFeatures<ST>*>(features);
+	index_t num_feat = dense_feat->get_num_features();
 
-	// covariance matrix.
-	auto cov_mat = linalg::matrix_prod(matrix, matrix, false, true);
-	SGMatrix<ST> scatter_matrix(num_feat, num_feat);
-	linalg::scale(cov_mat, scatter_matrix, 1 / (ST)(num_vec - 1));
+	auto solver = std::unique_ptr<CLDASolver<ST> >(new CLDASolver<ST>(
+		dense_feat,
+		new CMulticlassLabels(static_cast<CBinaryLabels*>(m_labels)),
```
Comment (karlnapf) [inline src/shogun/classifier/LDA.cpp]: same question as above

File: src/shogun/classifier/LDA.h [hunk 8]
```diff
@@ -82,6 +82,8 @@ template <class ST> class CDenseFeatures;
  *
  * <em>::AUTO_LDA</em> : This mode automagically chooses one of the above modes for
  * the users based on whether N > D (chooses ::FLD_LDA) or N < D(chooses ::SVD_LDA)
+ * Note that even if N > D FLD_LDA may fail being the covariance matrix not invertible,
```
Comment (karlnapf) [inline src/shogun/classifier/LDA.h]: What happens when this is the case? How is the user informed? Error message? Or automatic fallback?

File: src/shogun/classifier/LDA.h [hunk 9]
```diff
@@ -95,18 +97,22 @@ class CLDA : public CLinearMachine
 		 *
 		 * @param gamma gamma
 		 * @param method LDA using Fisher's algorithm or Singular Value Decomposition : ::FLD_LDA/::SVD_LDA/::AUTO_LDA[default]
+		 * @param accurate when using SVD solver switch between Jacobi's algorithm (more accurate)
```
Comment (karlnapf) [inline src/shogun/classifier/LDA.h]: is this described in the "@brief"?

File: src/shogun/labels/MulticlassLabels.cpp [hunk 10]
```diff
@@ -25,6 +25,14 @@ CMulticlassLabels::CMulticlassLabels(CFile* loader) : CDenseLabels(loader)
 	init();
 }
 
+CMulticlassLabels::CMulticlassLabels(CBinaryLabels* labels) : CDenseLabels(labels->get_num_labels())
+{
+	init();
+
+	for (index_t i = 0; i < labels->get_num_labels(); ++i)
```
Comment (karlnapf) [inline src/shogun/labels/MulticlassLabels.cpp]: nice!
@MikeLing this is what I had in mind as one of the things in label refactoring

File: src/shogun/labels/MulticlassLabels.h [hunk 11]
```diff
@@ -57,6 +57,14 @@ class CMulticlassLabels : public CDenseLabels
 		 */
 		CMulticlassLabels(CFile* loader);
 
+		/** constructor
```
Comment (karlnapf) [inline src/shogun/labels/MulticlassLabels.h]: Capital C in constructor. Also no need to say that it is constructor, so rather remove

File: src/shogun/classifier/LDA.cpp (hunk 2 above)
Comment (micmn) [inline src/shogun/classifier/LDA.cpp]: Added to train_machine

File: src/shogun/mathematics/linalg/LinalgEnums.h [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+#define LINALG_ENUMS_H__
+
+namespace shogun
+{
+
+namespace linalg
+{
+
+enum class SVDAlgorithm
+{
```
Comment (karlnapf) [inline src/shogun/mathematics/linalg/LinalgEnums.h]: The same thing could be done for other solvers, such as cholesky (LLT, LDLT, ...)

File: src/shogun/classifier/LDA.cpp (hunk 1 above)
Comment (micmn) [inline src/shogun/classifier/LDA.cpp]: The problem is that we cast ST to float64_t

File: src/shogun/mathematics/linalg/backend/eigen/Decompositions.cpp [hunk 13]
```diff
@@ -185,14 +186,37 @@ void LinalgBackendEigen::eigen_solver_symmetric_impl(
...
+		{
+			#if EIGEN_VERSION_AT_LEAST(3,3,0)
+			auto svd_eig =
+				A_eig.bdcSvd(thin_U ? Eigen::ComputeThinU : Eigen::ComputeFullU);
+			s_eig = svd_eig.singularValues().template cast<T>();
+			U_eig = svd_eig.matrixU().template cast<T>();
+			break;
+			#else
+			SG_SWARNING(
+				"At least Eigen 3.3 is required for BDC-SVD.\n"
```
Comment (karlnapf) [inline src/shogun/mathematics/linalg/backend/eigen/Decompositions.cpp]: no newline  `\n` after first sentence

File: src/shogun/classifier/LDA.cpp (hunk 5 above)
Comment (karlnapf) [inline src/shogun/classifier/LDA.cpp]: General question: Is this multiclass LDA? If so, why does it assert binary labels?
I think there should be no distinction (class wise) between binary and multiclass LDA

File: tests/unit/mathematics/linalg/operations/Eigen3_operations_unittest.cc [hunk 14]
```diff
@@ -1385,7 +1402,68 @@ TEST(LinalgBackendEigen, SGMatrix_svd_fullU)
...
+	SGMatrix<float64_t> U(m, n);
+	SGVector<float64_t> s(n);
+
+	svd(A, s, U, true, SVDAlgorithm::BidiagonalDivideConquer);
+
+	for (index_t i = 0; i < n; ++i)
+	{
+		auto c = CMath::sign(U[i * m] * result_U[i * m]);
+		for (index_t j = 0; j < m; ++j)
+			EXPECT_NEAR(U[i * m + j], c * result_U[i * m + j], 1e-7);
```
Comment (karlnapf) [inline tests/unit/mathematics/linalg/operations/Eigen3_operations_unittest.cc]: the indexing here seems a bit over-complicated

**Comment by vigsterkr** [comment]
@micmn there are some formatting errors: https://travis-ci.org/shogun-toolbox/shogun/jobs/250312428#L807

**Pull request merged**